enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,505 @@
+are updated in the `.status` of the corresponding wildcard `EgressFirewallDNSName` CRs.
+
+The following `EgressFirewallDNSName` CRD will be added to the `dns.openshift.io` api-group.
+
+````go
+// EgressFirewallDNSName describes a DNS name used in a EgressFirewall rule. It is TechPreviewNoUpgrade only.
+type EgressFirewallDNSName struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec is the specification of the desired behavior of the EgressFirewallDNSName.
+	Spec EgressFirewallDNSNameSpec `json:"spec,omitempty"`
+	// status is the most recently observed status of the EgressFirewallDNSName.
+	Status EgressFirewallDNSNameStatus `json:"status,omitempty"`
+}
+
+// EgressFirewallDNSNameSpec is a desired state description of EgressFirewallDNSName.
+type EgressFirewallDNSNameSpec struct {
+	// name is the DNS name used in a EgressFirewall rule.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.$
+	Name string `json:"name"`
+}
+
+// EgressFirewallDNSNameStatus defines the observed status of EgressFirewallDNSName.
+type EgressFirewallDNSNameStatus struct {
+	// The list of matching DNS names and their corresponding IP addresses along with TTL and last
+	// DNS lookup time.
+	ResolvedNames []EgressFirewallDNSNameStatusItem `json:"resolvedNames,omitempty"`
+}
+
+// EgressFirewallDNSNameStatusItem describes the details of a resolved DNS name.
+type EgressFirewallDNSNameStatusItem struct {
+	// The resolved DNS name corresponding to the Name field of EgressFirewallDNSNameSpec.
+	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.$
+	DNSName string `json:"dnsName"`
+	// The IP addresses associated with the DNS name used in a EgressFirewall rule.
+	IPs []string `json:"ips"`
+	// Minimum time-to-live value among all the IP addresses.
+	TTL int64 `json:"ttl"`
+	// Timestamp when the last DNS lookup was successfully completed.
+	LastLookupTime metav1.Time `json:"lastLookupTime"`
+}
+````
+
+### Workflow Description
+
+The workflows for Create, Delete and Update events for EgressFirewall related to DNS names are explained in this section. The workflow
+for those events is shown in the following diagram with an example `EgressFirewallDNSName` CR:
+
+![Workflow](./coredns-egressfirewall-integration.png)
+
+#### Create/Update of DNS name
+
+* An OpenShift cluster administrator creates/updates an EgressFirewall resource for a namespace and adds rule(s) containing
+DNS name(s).
+* The OVN-K cluster manager will create corresponding `EgressFirewallDNSName` CRs for each of the DNS names in the EgressFirewall rules, if not
+already created. Each CR will be created in the `openshift-ovn-kubernetes` namespace. The name of the CR will be assigned using a hash
```

Comment:
Crossing namespaces means you can't use the garbage collector built into K8s, why do we want all DNSNames in one namespace rather than keeping them in the same namespaces as the EgressFirewall that they belong to?

What data will form the input for the hash? Is this designed to deduplicate entries?

Reply:
> Crossing namespaces means you can't use the garbage collector built into K8s, why do we want all DNSNames in one namespace rather than keeping them in the same namespaces as the EgressFirewall that they belong to?

The DNS names are not mapped in an 1-1 fashion to the EgressFirewall objects. Same DNS name can be used in multiple EgressFirewall objects in different namespaces, however only a single EgressFirewallDNSName oject should be created for the DNS name. That's why it is created only in `openshift-ovn-kubernetes` namespace.

> What data will form the input for the hash? Is this designed to deduplicate entries?

The DNS name itself is used as the input for the hash as only a single CR should be created for each unique DNS name.

Reply:
This context isn't clear from the enhancement, please make sure that makes it to the document